### PR TITLE
ARROW-8779: [R] Implement conversion to List<Struct>

### PR DESCRIPTION
**Pull request by romainfrancois**
``` r
library(arrow)
#> 
#> Attaching package: 'arrow'
#> The following object is masked from 'package:utils':
#> 
#>     timestamp
Array$create(list(data.frame(a = 1:3, b = 11:13)))
#> ListArray
#> <list<item: struct<a: int32, b: int32>>>
#> [
#>   -- is_valid: all not null
#>   -- child 0 type: int32
#>     [
#>       1,
#>       2,
#>       3
#>     ]
#>   -- child 1 type: int32
#>     [
#>       11,
#>       12,
#>       13
#>     ]
#> ]
```

<sup>Created on 2020-06-15 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

However does not yet deal with the original issue: 

``` r
library(arrow)
#> 
#> Attaching package: 'arrow'
#> The following object is masked from 'package:utils':
#> 
#>     timestamp

nrows <- 1:3
df <- tibble::tibble(
    id = 1L,
    data = 
      list(
        tibble::tibble(
          a = letters[nrows],
          b = as.integer(nrows),
          c = as.factor(a),
          d = b/2
        )
      )
  )
Table$create(df)
#> Error in Table__from_dots(dots, schema): NotImplemented: Converting vector to arrow type dictionary<values=string, indices=int8, ordered=0> not implemented
```

<sup>Created on 2020-06-15 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

### CODE REVIEW

**Comment by romainfrancois** [comment]
@github-actions autotune everything

File: r/src/array_from_vector.cpp [hunk 1]
```diff
@@ -201,6 +201,39 @@ struct VectorToArrayConverter {
...
+
+    int64_t n = 0;
+    if (XLENGTH(x) > 0) {
+      // TODO: need a more generic way to get the vec_size(<data.frame>)
+      //       perhaps using vctrs::::short_vec_size
+      n = XLENGTH(VECTOR_ELT(x, 0));
+    }
+    RETURN_NOT_OK(struct_builder->Reserve(n));
+
+    // TODO: double check but seems fine that no value is NULL
```
Comment (fsaintjacques) [inline r/src/array_from_vector.cpp:220]: You can remove the TODO, this is correct.

**Review by fsaintjacques — CHANGES_REQUESTED** [review]

**Comment by fsaintjacques** [comment]
I also updated ARROW-7798 with comments to avoid builder in the conversion.

File: r/tests/testthat/test-Array.R [hunk 2]
```diff
@@ -445,6 +445,9 @@ test_that("Array$create() handles vector -> list arrays (ARROW-7662)", {
   expect_array_roundtrip(list(character(0)), list_of(utf8()))
   expect_array_roundtrip(list("itsy", c("bitsy", "spider"), c("is")), list_of(utf8()))
   expect_array_roundtrip(list("itsy", character(0), c("bitsy", "spider", NA_character_), c("is")), list_of(utf8()))
+
+  # struct
+  expect_array_roundtrip(list(tibble::tibble(a = integer(0))), list_of(struct(a = int32())))
```
[the remark this answers is not in the document]
Comment (romainfrancois) [reply r/tests/testthat/test-Array.R:450]: Thanks.

**Comment by romainfrancois** [comment]
I need to add tests for this: 

``` r
library(arrow, warn.conflicts = FALSE)

nrows <- 1:3
df <- tibble::tibble(
  id = 1L,
  data = 
    list(
      tibble::tibble(
        a = letters[nrows],
        b = as.integer(nrows),
        c = as.factor(a),
        d = b/2
      )
    )
)
tab <- Table$create(df)
tab$schema
#> Schema
#> id: int32
#> data: list<item: struct<a: string, b: int32, c: dictionary<values=string, indices=int8>, d: double>>
as.data.frame(tab)
#> # A tibble: 1 x 2
#>      id data            
#>   <int> <list>          
#> 1     1 <tibble [3 × 4]>
as.data.frame(tab)$data
#> [[1]]
#> # A tibble: 3 x 4
#>   a         b c         d
#>   <chr> <int> <fct> <dbl>
#> 1 a         1 a       0.5
#> 2 b         2 b       1  
#> 3 c         3 c       1.5
```

<sup>Created on 2020-06-16 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by romainfrancois** [comment]
also added `VectorToArrayConverter::Visit` for dictionaries to that we can handle things like list(factor()), the implementation is simpler than the original `MakeFactorArray` as it leverages `StringDictionaryBuilder`, but it looks like that does not respect the "ordered" property. 

``` r
library(arrow, warn.conflicts = FALSE)

o <- ordered(c("b", "a"), levels = c("a", "b", "c"))

# this goes trough MakeFactorArray
Array$create(o)$type
#> DictionaryType
#> dictionary<values=string, indices=int8, ordered>

# this goes through VectorToArrayConverter::Visit >> MakeBuilder() ...
Array$create(list(o))$type$value_type
#> DictionaryType
#> dictionary<values=string, indices=int8>
```

<sup>Created on 2020-06-16 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

File: r/src/array_from_vector.cpp [hunk 3]
```diff
@@ -201,6 +202,67 @@ struct VectorToArrayConverter {
...
 
+  template <typename T>
+  arrow::enable_if_t<is_struct_type<T>::value, Status> Visit(const T& type) {
+    using BuilderType = typename TypeTraits<T>::BuilderType;
+    ARROW_RETURN_IF(!Rf_inherits(x, "data.frame"),
+                    Status::RError("Expecting a data frame"));
+
+    auto* struct_builder = checked_cast<BuilderType*>(builder);
+
+    int64_t n = vctrs::short_vec_size(x);
```
Comment (fsaintjacques) [inline r/src/array_from_vector.cpp:213]: Why do you need this?

**Review by fsaintjacques — APPROVED** [review]

File: r/src/array_from_vector.cpp (hunk 3 above)
Comment (wesm) [reply r/src/array_from_vector.cpp:213]: It seems in R it's possible to have data-frame-like tibble columns that are malformed viewed as Arrow-like (as evidenced by the unit test)

**Comment by wesm** [comment]
thanks @romainfrancois!

File: r/src/array_from_vector.cpp (hunk 3 above)
Comment (romainfrancois) [reply r/src/array_from_vector.cpp:213]: `vctrs::short_vec_size()` knows how to get the size of things like data frames with 0 columns, data frames with data frame columns ...

Comment (nealrichardson) [reply r/src/array_from_vector.cpp:213]: See also https://jira.apache.org/jira/browse/ARROW-9028 and some skipped tests about tables with 0 columns

**Pull request closed without merging**